include/deal.II/lac/petsc_snes.h

```diff
@@ -0,0 +1,477 @@
+
+    /**
+     * Initialization parameters for NonlinearSolverData.
+     *
+     * Running parameters:
+     *
+     * @param opts_prefix The string indicating the options prefix for command line customization.
+     * @param snestype The string indicating the PETSc SNES solver type.
+     * @param sneslstype The string indicating the PETSc linesearch type.
+     * @param absolute_tolerance Absolute error tolerance.
+     * @param relative_tolerance Relative error tolerance.
+     * @param step_tolerance Step tolerance.
+     * @param max_it Maximum number of iterations allowed.
+     * @param max_fe Maximum number of function evaluations allowed.
+     *
+     * Note that all parameters values specified here can be overriden by
+     * command line choices.
+     */
+    NonlinearSolverData(
+      // Running parameters
+      const std::string &opts_prefix        = "",
+      const std::string &snestype           = "",
+      const std::string &sneslstype         = "",
+      const real_type    absolute_tolerance = 0,
+      const real_type    relative_tolerance = 0,
+      const real_type    step_tolerance     = 0,
+      const int          max_it             = -1,
+      const int          max_fe             = -1)
+      : opts_prefix(opts_prefix)
+      , snestype(snestype)
+      , sneslstype(sneslstype)
+      , absolute_tolerance(absolute_tolerance)
+      , relative_tolerance(relative_tolerance)
+      , step_tolerance(step_tolerance)
+      , max_it(max_it)
+      , max_fe(max_fe)
+    {}
+
+    void
+    add_parameters(ParameterHandler &prm)
+    {
+      prm.enter_subsection("Running parameters");
+      prm.add_parameter(
+        "options prefix",
+        opts_prefix,
+        "The string indicating the options prefix for command line customization.");
+      prm.add_parameter("solver type",
+                        snestype,
+                        "The string indicating the PETSc SNES type.");
+      prm.add_parameter("linesearch type",
+                        sneslstype,
+                        "The string indicating the PETSc linesearch type.");
+      prm.add_parameter("absolute error tolerance",
+                        absolute_tolerance,
+                        "Absolute error tolerance.");
+      prm.add_parameter("relative error tolerance",
+                        relative_tolerance,
+                        "Relative error tolerance.");
+      prm.add_parameter("step tolerance", step_tolerance, "Step tolerance.");
+      prm.add_parameter("maximum iterations",
+                        max_it,
+                        "Maximum number of iterations allowed.");
+      prm.add_parameter("maximum function evaluations",
+                        max_fe,
+                        "Maximum number of function evaluations allowed.");
+      prm.leave_subsection();
+    }
+
+    /**
+     * Options database prefix.
+     */
+    std::string opts_prefix;
+
+    /**
+     * PETSc solver type.
+     */
+    std::string snestype;
```

Comment:
```suggestion
    std::string snes_type;
```

Reply:
Would it make sense to introduce enums?

Reply:
no, because PETSc types are dynamic, in the sense that users can register new ones, and the way we associate types to implementation is through strings.